### ISSUE TITLE
Blazor wasm hosted Project Integrity errors on Deployment to Azure App service

### ISSUE THREAD

**Issue by nssidhu** [dotnet/aspnetcore#37861]
This is .NET 6 RC2 Blazor Hosted WebAssembly App and deploying it to Azure App service (windows)
I am getting lots of integrity related errors .

I tired running locally in Release Mode and browsing it in Incognito mode without problem.

```
index.html:1 
        
       Failed to find a valid digest in the 'integrity' attribute for resource 'https://getinlinedev-staging.azurewebsites.net/_framework/Serilog.Enrichers.Environment.dll' with computed SHA-256 integrity 'Lbxblrss7XKRuEZFwDckN1swP9G9Qpa7WAHFILwpw8c='. The resource has been blocked.
```

some more errors

```
Uncaught (in promise) TypeError: Failed to fetch
    at lt.loadResourceWithoutCaching (blazor.webassembly.js:1)
    at lt.loadResourceWithCaching (blazor.webassembly.js:1)
```

```
index.html:1 
        
       Failed to find a valid digest in the 'integrity' attribute for resource 'https://getinlinedev-staging.azurewebsites.net/_framework/dotnet.timezones.blat' with computed SHA-256 integrity 'IRA3aSR05YjbbkfXMY8trRnmm9TfDlmrKiB4EPRl16I='. The resource has been blocked.


Failed to find a valid digest in the 'integrity' attribute for resource 'https://getinlinedev-staging.azurewebsites.net/_framework/dotnet.wasm' with computed SHA-256 integrity '1Qbl2oLjBqIKta0l6OQZQH03S0ZS/S9M18GsKlAnhuQ='. The resource has been blocked.
index.html:1 
        
       Unknown error occurred while trying to verify integrity.
```

```
service-worker.js:37 
        
       Uncaught (in promise) TypeError: Failed to fetch
    at service-worker.js:37
    at async onInstall (service-worker.js:37)
```

this is how my service-wroker.published.js looks

```
// Caution! Be sure you understand the caveats before publishing an application with
// offline support. See https://aka.ms/blazor-offline-considerations

self.importScripts('./service-worker-assets.js');
self.addEventListener('install', event => event.waitUntil(onInstall(event)));
self.addEventListener('activate', event => event.waitUntil(onActivate(event)));
self.addEventListener('fetch', event => event.respondWith(onFetch(event)));

const cacheNamePrefix = 'offline-cache-';
const cacheName = `${cacheNamePrefix}${self.assetsManifest.version}`;
const offlineAssetsInclude = [ /\.dll$/, /\.pdb$/, /\.wasm/, /\.html/, /\.js$/, /\.json$/, /\.css$/, /\.woff$/, /\.png$/, /\.jpe?g$/, /\.gif$/, /\.ico$/, /\.blat$/, /\.dat$/ ];
const offlineAssetsExclude = [ /^service-worker\.js$/ ];

const notifyNewVersion = () => {
    const bc = new BroadcastChannel('blazor-channel');

    bc.postMessage('new-version-found');

    bc.onmessage = function (message) {
        if (message && message.data == "skip-waiting") {
            console.info("Calling skipWaiting");
            self.skipWaiting();
            bc.postMessage("reload-page");
        }
    }
}

async function onInstall(event) {
    console.info('Service worker: Install');

    // Fetch and cache all matching items from the assets manifest
    const assetsRequests = self.assetsManifest.assets
        .filter(asset => offlineAssetsInclude.some(pattern => pattern.test(asset.url)))
        .filter(asset => !offlineAssetsExclude.some(pattern => pattern.test(asset.url)))
       // .map(asset => new Request(asset.url, { integrity: asset.hash }));
        .map(asset => new Request(asset.url, { integrity: asset.hash, cache: 'no-cache' }));
    await caches.open(cacheName).then(cache => cache.addAll(assetsRequests));
}

async function onActivate(event) {
    console.info('Service worker: Activate');

    // Delete unused caches
    const cacheKeys = await caches.keys();
    await Promise.all(cacheKeys
        .filter(key => key.startsWith(cacheNamePrefix) && key !== cacheName)
        .map(key => caches.delete(key)));
}

async function onFetch(event) {
    let cachedResponse = null;
    if (event.request.method === 'GET') {
        // For all navigation requests, try to serve index.html from cache
        // If you need some URLs to be server-rendered, edit the following check to exclude those URLs
        const shouldServeIndexHtml = event.request.mode === 'navigate';

        const request = shouldServeIndexHtml ? 'index.html' : event.request;
        const cache = await caches.open(cacheName);
        cachedResponse = await cache.match(request);
    }

    return cachedResponse || fetch(event.request);
}

```

**Comment by JasonFreeberg** [comment]
Hey @nssidhu , I'm on App Service. Wanted to confirm a couple things:

- What client are you deploying from? Visual Studio? What version?
- You're deploying to a Windows web app, correct?

**Comment by nssidhu** [comment]
Visual Studio 2022 Version 17.0.0 Preview 6.0.
Deploying to Azure App Service Windows, as Web App

**Comment by nssidhu** [comment]
I have again run into the integrity error. 
This time, the only thing I did was, I upgraded from .NET RC2 to .NET 6 version.

![image](https://user-images.githubusercontent.com/3267605/141350923-02752802-7fa9-4bde-b96d-5a9cb6717fbd.png)

Here is how i am publishing from Visual Studio 2022. The publish had been working fine, it just broke after upgrading to .NET 6
![image](https://user-images.githubusercontent.com/3267605/141351073-5c2133b5-cc1d-4122-917c-a4eac01ad25e.png)

In case someone want to inspect or have a look, I will keep everything intact for a day, before I start to manually clean the App service environment.

**Comment by nssidhu** [comment]
some interesting findings

I incorrectly assumed that something is wrong on the Azure App Service environment.

It turns out that manually deleting the obj & bin folders for the projects helps.

This is what i tried
1) Published my project to local folder and tried to run that from the published folder.
   This gave me exact error that i would get if published on Azure app Service.
   this itself proves that issue is not related to Azure App Service Environment.
   
2) I manually deleted the published local folder to prepare for new publish.
   second i also deleted all bin & obj folders from my project.
   I again published to local folder and tried to run the app from local folder.
   This time i was able to run it without any problem.
 
3) I Published the project to Azure App Service.(I had checked "Remove Additional files at destination"
   I was able to browse the website without any kind of integrity errors.
   
   Hope this helps save day for someone who might run into this problem.
   
   So it turns out that Clean & rebuild in Visual Studio 2022 is not able to remove some of the cached files.

**Comment by nssidhu** [comment]
@javiercn  Can this issue be posted/reported to Visual Studio team so that they can enable the deleting of bin & obj folders on publish?  NOTE: They do have clean option but that itself does not help, may the clean might have some bugs and does not work as expected.

**Comment by javiercn** [comment]
@nssidhu thanks for contacting us.

Rebuild is just a combination of Clean + Build so if there's a clean option, that should work. 

What seems to be happening, from what I understand, is that you are running into issues when publishing more than once? (Even after cleaning in between).

Does this happen when you publish, delete the publish directory and publish again?

**Comment by nssidhu** [comment]
Agree, but looks like Clean or Rebuild both does not work as intended.
Yes I did delete the publish directory before republishing, when run it produces the same error.

The only way out is manually deleting the bin & obj folder from file explorer and than doing the builds or directly publish(which builds it anyway).

I recently upgraded the project to .NET 6 LTS, which resulted into integrity error.
So rather than multiple cycles of publish(which I thought earlier might have contributed to this), looks like it occurs when some major package changes are done to the project.
I my recent issue I just updated my project to .NET 6 LTS once only and integrity errors started showing up.

**Comment by javiercn** [comment]
@nssidhu thanks for the additional details.

Does the issue disappear once you've updated the project and cleaned the bin and obj folders? (Meaning that after the upgrade and removing the folders, the sequence publish, delete publish folder, publish) works again?

**Comment by nssidhu** [comment]
Yes


Sent from Yahoo Mail for iPhone


On Friday, November 12, 2021, 9:00 AM, Javier Calvarro Nelson ***@***.***> wrote:




@nssidhu thanks for the additional details.

Does the issue disappear once you've updated the project and cleaned the bin and obj folders? (Meaning that after the upgrade and removing the folders, the sequence publish, delete publish folder, publish) works again?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by javiercn** [comment]
@nssidhu thanks for the additional details.

Seems then that the issue only happens when updating a project between 5.0 and 6.0.

**Comment by nssidhu** [comment]
That was the most recent one, it does happens when there are major package changes.
I remember another instance when I added some globalization package I ran into integrity issue there too. I have tagged you on that issue but for different issue.


Sent from Yahoo Mail for iPhone


On Friday, November 12, 2021, 10:05 AM, Javier Calvarro Nelson ***@***.***> wrote:




@nssidhu thanks for the additional details.

Seems then that the issue only happens when updating a project between 5.0 and 6.0.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by javiercn** [comment]
@nssidhu thanks for the additional details.

Our current guidance is to clean up the project artifacts when making major upgrades as the complexity of getting the upgrade to work in a non clean environment is very high.

**Comment by WayneHiller** [comment]
I had the same issue if the app was copied to any other machine I would get errors. On my machine it would run fine from the Publish folder. I did not change the dotnet 6 version (Release). Deleting the bin + obj folders fixed it. 

Interestingly I verified the dll version and dates reported in the errors were exactly the same as my publish folder. If I checked the Hash of one on my machine it was the same as listed in the service worker assents file. However if I checked the Hash on the other machine it was different. Same issue deploying to Ubuntu in a Docker container.
